### ISSUE TITLE
difficulty generalizing examples

### ISSUE THREAD

**Issue by micahscopes** [rlabbe/filterpy#69]
I'm really loving your book and library 👍 

I just wanted to comment that I'm having some trouble making the jump from the examples given in the book to applying tools from the filterpy library.  Specifically, I've been reading the chapter on adaptive filtering and tracking, and I'm finding that it's not always easy for me to follow what's going on in the examples.  There is a lot of copy/pasted code and functions that are either redefined a lot or not used from one cell to the next... then, suddenly they're used again.  

Don't get me wrong, I'm learning a lot from this chapter!  But I tried lifting in my own data by adapting the `generate_data` function to spit out data formatted in the exact same way as the sample data, and it just didn't work out very well.  The `initialize_filter` function makes the assumption that the first data point is going to be 0, which wasn't the case for my data, and my data was a time series, which is technically two dimensional, but it didn't fit well with the way that the filters are being applied.

I think the main thing is that at the moment is that the book is the best documentation for some of the more adaptive filtering tools in the filterpy library, and it'd be nice to have some more generic working examples for how to use the IMM and MMAE tools.

Thanks a lot for doing this... It's really helpful!

**Comment by micahscopes** [comment]
By the way, the whole reason I tried going that route in the first place was that I had some funny errors when trying to use the IMMEstimator with two different filter objects:

```python
import copy
from filterpy.kalman import IMMEstimator
from scipy.linalg import block_diag

# N = 600
dt = 0.0001
std = 0.05

ca = KalmanFilter(dim_x=3, dim_z=1)
ca.x = np.array([0., 0., 0.])
ca.P *= 3
ca.R *= std
ca.Q = Q_discrete_white_noise(dim=3, dt=dt, var=0.02)
ca.F = np.array([[1, dt, 0.5*dt*dt],
                       [0, 1,         dt], 
                       [0, 0,          1]])
ca.H = np.array([[1., 0, 0]])

cano = KalmanFilter(dim_x=3, dim_z=1)
cano.x = np.array([0., 0., 0.])
cano.P *= 3
cano.R *= std
cano.Q = Q_discrete_white_noise(dim=3, dt=dt, var=0.02)
cano.F = np.array([[1, dt, 0.5*dt*dt],
                       [0, 1,         dt], 
                       [0, 0,          1]])
cano.H = np.array([[1., 0, 0]])

filters = [ca, cano]



M = np.array([[0.8, 0.2],
              [0.2, 0.8]])
mu = np.array([0.5, 0.5])
bank = IMMEstimator(filters, mu, M)


zs = dat.H.values
xs, probs = [], []
cvxs, caxs = [], []
for i, z in enumerate(zs):
    z = np.array([z]).T
    bank.update(z)

    xs.append(bank.x.copy())

xs = np.array(xs)

plt.plot(dat.C.values)
plt.plot(xs[:,0])
plt.show()
```
This gives me:
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-130-09baf6f9d877> in <module>()
     42 for i, z in enumerate(zs):
     43     z = np.array([z]).T
---> 44     bank.update(z)
     45 
     46     xs.append(bank.x.copy())

~/.local/lib/python3.6/site-packages/filterpy/kalman/IMM.py in update(self, z, u)
     93         L = zeros(len(self.filters))
     94         for i, f in enumerate(self.filters):
---> 95             f.update(z)
     96             L[i] = f.likelihood
     97 

~/.local/lib/python3.6/site-packages/filterpy/kalman/kalman_filter.py in update(self, z, R, H)
    186         # K = PH'inv(S)
    187         # map system uncertainty into kalman gain
--> 188         self.K = dot3(P, H.T, linalg.inv(self.S))
    189 
    190         # x = x + Ky

~/.local/lib/python3.6/site-packages/scipy/linalg/basic.py in inv(a, overwrite_a, check_finite)
    788 
    789     """
--> 790     a1 = _asarray_validated(a, check_finite=check_finite)
    791     if len(a1.shape) != 2 or a1.shape[0] != a1.shape[1]:
    792         raise ValueError('expected square matrix')

~/.local/lib/python3.6/site-packages/scipy/_lib/_util.py in _asarray_validated(a, check_finite, sparse_ok, objects_ok, mask_ok, as_inexact)
    236             raise ValueError('masked arrays are not supported')
    237     toarray = np.asarray_chkfinite if check_finite else np.asarray
--> 238     a = toarray(a)
    239     if not objects_ok:
    240         if a.dtype is np.dtype('O'):

/usr/lib/python3.6/site-packages/numpy/lib/function_base.py in asarray_chkfinite(a, dtype, order)
   1213     if a.dtype.char in typecodes['AllFloat'] and not np.isfinite(a).all():
   1214         raise ValueError(
-> 1215             "array must not contain infs or NaNs")
   1216     return a
   1217 

ValueError: array must not contain infs or NaNs
```

**Comment by rlabbe** [comment]
Okay, I played with it a bit, I see the problem. Mind if I add this as a section in the book for debugging?

Since I don't have your data, I placed the following in the code, and sure enough, got the exception. 

N=100
zs = np.array([x for x in range(N)], dtype=float)

Why? I looked at the plots, and the bank is not tracking the data well. I added a print of S and P right after the update call and saw that cano.P  had all nan's in it. This will happen when the KF diverges. KF will diverge if they are not designed well.

So I ran this code - just the cano filter:

for i, z in enumerate(zs):
    z = np.array([z]).T
    cano.update(z)
    print(i, z, cano.P)
    cano.predict()
    xs.append(cano.x.copy())

And while the filter works, it diverges from the data. But if I change N to 10000, it eventually converges again. 

dt is very small, so N=100 is a tiny bit of time: 1/00 sec. 

I changed the initialization of x to np.array([0., 10000., 0.]). Which is really just [0, v/dt, 0] where v/epoch=1.0.

Now the filter tracks fine and doesn't crash. 

Another way is to realize that you are really over certain in your definition of P. You have P=diag([3,3,3]), yet the variance is really far greater - you are giving an initial velocity of 0, but  the actual velocity is 10000.

Of course, this is my data, not yours, you may be encountering a different problem, but one way or another I suspect your filter is not well designed for your data.

I'm not entirely sure how the IMM code is interacting with this. Presumably there is a divide by zero somewhere. I can look at that later. This is actually a really good illustration of the difference between pedagogical code and production code. If you just code the equations from the books you get something that is theoretically correct yet works poorly in practice Or, I just have a bug in my IMM code ;). 

Basically, though, you have to be careful with initializing x; I should stress that more. It may not 'matter' when running a single KF, except the output is poor, but once you start trying to use that information in other ways all heck can break loose because you do not have a well performing filter. In practice I write code that initialize v and a after the 2 and 3rd update if running real time, and pre-compute it if running in batch mode. I.e. v_0 = z[1] - z[0] if z is in state space.

**Comment by rlabbe** [comment]
Okay, more debugging.... looking in the IMM code there are two divisions, the divide by cbar, and then the normalization of mu. Turns out the latter generates nan, which then propagates into P in subsequent iterations.


```python
        self.mu =  self.cbar * L
        self.mu /= sum(self.mu) # normalize
```
L is the likelihood of each filter. On inspection it is 0 for both. So essentially the filter has diverged so far that the math for the likelihood (which is the probability that the measurement fits the filter) ends up being 0. I'm sure this is a finite precision issue - it is probably 10^343234 or some ridiculously small number. 

If I add this code after bank.update() my filter runs without issue (again, with my data, not yours):
```python
    if not np.isscalar(bank.mu):
        bank.mu = np.array([0.5, 0.5])
```

**Comment by rlabbe** [comment]
Okay, hopefully my final message. In my code, cano.log_likelihood = -855.3628. exp() of that = 0, so likelihood = 0. We can constrain kalman_filter.log_likelihood >= np.log(np.finfo(float).tiny), ensuring likelihood is always > 0. that will require a code change to kalman_filter.py

**Comment by micahscopes** [comment]
Wow thank you so much for your response!  Feel free to use this as a
debugging example.

As for the precision issue... Very interesting... Reminds me of a hack I
made for the MMAE method to keep the probabilities from converging to 0 or
1.  I  placed a limit on how close they could get to 0 or 1.... It sorta
worked, but wasn't easy to control!

I also noticed that the shapes of the constant acceleration and constant
velocity filters are different, and this seemed to be a problem for the
IMMEstimator's assumptions about updating. I was wondering if it'd be
possible to extend this class, or the filters themselves, to allow passing
in custom lambdas or functions for each filter which specify how new
updates should map... (This is still a basic misunderstanding of mine, as
I'm still wrapping my head around how the constant velocity/acc filters
work)

By the way, I made some "constant jerk" and "constant snap" filters too!
I'm curious to incorporate these with the IMM, or even just to specify a
list of orders you want the model to fix as constant. My reasoning was that
I wanted to switch back and forth between nearly constant velocity and very
quick turns. And just because I was curious.

Since I last posted, i realized some "gotchas". For one, because of how my
data was resampled, there were long stretches of repeated values even when
their shouldn't have been any update. I made a change in my update logic to
pass None when a new observation was the same as the last one.  This helped
my filters fit the data better.

Okay one last thought... Have you considered using pandas for the book? I
just love it. It could really help clarify what's going on with certain
examples. Not that there's anything wrong with numpy... But for educational
material I think pandas could help reduce cognitive overhead!
